### PR TITLE
Speed up aggs with sub-aggregations

### PR DESCRIPTION
**Pull request by nik9000**
This allows many of the optimizations added in #63643 and #68871 to run
on aggregations with sub-aggregations. This should:
* Speed up `terms` aggregations on fields with less than 1000 values that
  also have sub-aggregations. Locally I see 1.5 second searches run in 1.2
  seconds.
* Applies that same speedup to `range` and `date_histogram` aggregations but
  it feels less impressive because the point range queries are a little
  slower to get up and go.
* Massively speed up `filters` aggregations with sub-aggregations that
  don't have a `parent` aggregation or collect "other" buckets. Also
  save a ton of memory while collecting them. I've seen reports of this
  being a 20x speed up. Even if `filters` is rare that's pretty darn good.

### CODE REVIEW

**Comment by nik9000** [comment]
I've opened this up as a draft so I can use jenkins to run all the tests in parallel. My desktop is busy at the moment....

**Comment by nik9000** [comment]
```
| 90th percentile service time |                     keyword-terms | 1666.85   | 1555.73  | -111.116   | ms |
| 90th percentile service time |     keyword-terms-low-cardinality |   37.6112 |   38.016 |    0.40475 | ms |
| 90th percentile service time |                 keyword-terms-min | 2875.04   | 2934.54  |   59.493   | ms |
| 90th percentile service time | keyword-terms-low-cardinality-min | 1515.4    | 1151.45  | -363.957   | ms |
```

Those are the rally reslts for this change. The 24% speed up on
`keyword-terms-low-cardinality-min` is expacted. The speed up on
`keyword-terms` is not. Gremlins? Either way, this is
`keyword-terms-low-cardinality-min`:

```
POST weather-data-2016/_search
{
  "size": 0,
  "aggs": {
    "country": {
      "terms": {
        "field": "station.country",
        "size": 200
      },
      "aggs": {
        "tmin": {
          "min": {
            "field": "TMIN"
          }
        }
      }
    }
  }
}
```

**Comment by nik9000** [comment]
An interesting thing - this totally ignores `collect_mode` right now. If you get into the "less than 1000 distinct values" branch we'll use `filters` agg and just collect immediately. This is maybe ok because usually you'll want an appreciable portion of those 1000 distinct values anyway. But it is probably worth revisiting `collect_mode` in light of this.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java [hunk 1]
```diff
@@ -135,8 +135,22 @@ public static FiltersAggregator build(
...
+                keyed,
+                otherBucketKey,
+                context,
+                parent,
+                cardinality,
+                metadata
+            );
+            if (false == filterByFilter.scoreMode().needsScores()) {
+                // Filter by filter won't produce the correct results if the sub-aggregators need scores
+                return filterByFilter;
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java:152]: This bit is pretty sad! But it isn't worth trying to figure this stuff out now. Maybe not never. I wish it weren't so janky to stop it.

**Comment by nik9000** [comment]
I think I have the tests sorted. This brings a 24% speed up when a low cardinality `terms` aggregation is followed by a simple metric aggregation. And it makes it possible for us to enhance things like `terms -> terms -> metrics` and `date_histogram -> terms -> metrics` which are both fairly common.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/QueryToFilterAdapter.java [hunk 2]
```diff
@@ -216,12 +217,31 @@ long count(LeafReaderContext ctx, FiltersAggregator.Counter counter, Bits live)
...
+    /**
+     * Estimate the cost of calling {@code #count} in a leaf.
+     */
+    long estimateCollectCost(LeafReaderContext ctx) throws IOException {
         BulkScorer scorer = bulkScorer(ctx, () -> scorersPreparedWhileEstimatingCost++);
         if (scorer == null) {
             // There aren't any matches for this filter in this leaf
             return 0;
         }
-        return scorer.cost();   // TODO in another PR (please) change this to ScorerSupplier.cost
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/QueryToFilterAdapter.java:224]: Why did we lose this TODO?

**Review by not-napoleon — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/QueryToFilterAdapter.java (hunk 2 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/QueryToFilterAdapter.java:224]: I *thouhgt* I just removed the "in another PR" part. Let me fix.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java (hunk 1 above)
Comment (imotov) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java:152]: Yeah, that scares me quite a bit. Can you add a comment explaining that this information is not readily available until we actually build aggregations. I feel like factories should have enough information to actually know this stuff.

**Review by imotov — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java (hunk 1 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java:152]: They might be able to tell, but it'd be a ton of plumbing. I looked at it and though "I think it'd be easier to make score work here". But I didn't want to do either one right now..... I'll add the comment!

Comment (imotov) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java:158]: Yeah, I looked at it as well, and it didn't look practical to fix right now. My comment was more like a mental note to add this to the wish list if we ever get to refactor them.

**Comment by nik9000** [comment]
OK! Backport in. Time to update skips and such in master.

**Pull request merged**